dashboard/modules/job/rayletless_job_head.py

```diff
@@ -0,0 +1,399 @@
+        """
+        super().__init__(
+            address=dashboard_agent_address,
+            create_cluster_if_needed=create_cluster_if_needed,
+            cookies=cookies,
+            metadata=metadata,
+            headers=headers,
+        )
+
+    def submit_job_internal(
+        self,
+        *,
+        entrypoint: str,
+        job_id: Optional[str] = None,
+        runtime_env: Optional[Dict[str, Any]] = None,
+        metadata: Optional[Dict[str, str]] = None,
+        submission_id: Optional[str] = None,
+    ) -> JobSubmitResponse:
+        if job_id:
+            logger.warning(
+                "job_id kwarg is deprecated. Please use submission_id instead."
+            )
+
+        runtime_env = runtime_env or {}
+        metadata = metadata or {}
+        metadata.update(self._default_metadata)
+
+        # Run the RuntimeEnv constructor to parse local pip/conda requirements files.
+        runtime_env = RuntimeEnv(**runtime_env).to_dict()
+
+        submission_id = submission_id or job_id
+
+        req = JobSubmitRequest(
+            entrypoint=entrypoint,
+            submission_id=submission_id,
+            runtime_env=runtime_env,
+            metadata=metadata,
+        )
+
+        logger.debug(f"Submitting job with submission_id={submission_id}.")
+
+        r = self._do_request(
+            "POST",
+            "/api/job_agent/jobs/",
+            json_data=dataclasses.asdict(req),
+        )
+
+        if r.status_code == 200:
+            return JobSubmitResponse(**r.json())
+        else:
+            self._raise_error(r)
+
+
+# TODO(Catch-Bull): because of `routes` will check duplicate route path,
+# So `RayletlessJobHead` cannot inherit `JobHead`, Many functions are
+# copied directly from `JobHead`. Finally, after RayletlessJobHead is stable,
+# we will delete `JobHead`.
+class RayletlessJobHead(dashboard_utils.DashboardHeadModule):
+    def __init__(self, dashboard_head):
+        super().__init__(dashboard_head)
+        self._dashboard_head = dashboard_head
+        self._gcs_job_info_stub = None
+        self._job_info_store_client = None
+
+        # this is a queue of JobAgentSubmissionClient
+        self._agents = None
+
+    async def choose_agent(self) -> Optional[str]:
+        """
+        Try to disperse as much as possible to select one of
+        the `CANDIDATE_AGENT_NUMBER` agents to solve requests.
+        the agents will not pop from `self._agents` unless
+        it's dead.
+
+        Follow the steps below to select the agent client:
+            1. delete dead agent from `self._agents`, make sure
+               the `JobAgentSubmissionClient` in `self._agents`
+               is always available.
+            2. Attempt to put new agents into `self._agents` until
+               its size is `CANDIDATE_AGENT_NUMBER`
+            3. Returns the element at the head of the `self._agents`
+               and put it into `self._agents` again.
+        """
```

Comment:
Should we break out all of this queueing logic into a class and unit test it?  Also, what's the purpose of `CANDIDATE_AGENT_NUMBER`?  When should it be set to 1 and when should it be set to something higher?

Reply:
ok, I will add a relevant test.
`CANDIDATE_AGENT_NUMBER ` is the number of `JobAgents` involved in submitting jobs. The reason for this is to control the number of `JobAgents` that will call `ray.init`

Reply:
I see, I understand the low-level meaning of `CANDIDATE_AGENT_NUMBER` but I think I'm still missing the reason why we need to control this number.  From the user's perspective, what are the consequences of it being too high, and what are the consequences of it being too low?